sw/device/silicon_creator/rom/doc/bootstrap.md

```diff
@@ -0,0 +1,193 @@
+
+*   Upon entering bootstrap mode, the ROM will initialize EEPROM discovery mechanisms, such as the JEDEC ID and Serial Flash Discovery Parameters (SFDP) Table.
+*   Bootstrap will wait for an ERASE opcode; all other operations will be forbidden.
+    *   READs will return a static buffer of 0xFF.
+    *   PROGRAM operations will do nothing.
+*   Upon receiving an ERASE opcode, bootstrap will erase the internal flash data partitions and then enter a generic SPI flash opcode dispatch loop.
+The erase operation will not affect any of the flash info partitions.
+*   Within the generic opcode dispatch loop:
+    *   An ERASE opcode will validate the supplied address and erase the target sector.
+    *   A PAGE\_PROGRAM opcode will validate the supplied address and program the target page.
+    *   A RESET opcode will exit the dispatch loop and reset the chip.
+
+![Bootstrap Programming Flow](bootstrap_flows.svg)
+
+**Figure 1: Simplified EEPROM-mode Bootstrap Flow**
+
+## Implementation
+
+### Public API
+
+```
+// Determine whether or not bootstrap mode has been requested by
+// examining the strapping pins configuration.
+bool bootstrap_mode_check(void);
+
+// Enter into bootstrap mode and perform the update protocol.
+// - Initialize the SPI Device
+// - Configure SPI JEDEC ID and SFDP tables.
+// - Listen and act on SPI transactions
+rom_error_t bootstrap_mode_enter(void);
+```
+
+### Internal API & Pseudo Code
+
+```
+// Set the jedec ID in the SPI device
+void bootstrap_jedecid_set(uint8_t *id, size_t len);
+
+// Set the SFDP table in the SPI device.
+// TODO: define sfdp_t.
+void bootstrap_sfdp_set(sfdp_t *sfdp);
+
+// Perform the initial bootstrap phase:
+// Wait for an ERASE opcode, and silently ack all others.
+// Perform the ERASE when received.
+// Returns OK after the ERASE completes.
+rom_error_t bootstrap_phase1(void)
+
+// Perform the main bootstrap phase
+// Implement ERASE, PROGRAM and RESET opcodes, and silently ack all others.
+// We expect proper termination of the bootstrap protocol to include
+// a RESET opcode.
+// Returns OK after receiving the RESET opcode.
+rom_error_t bootstrap_phase2(void);
+
+// Ask reset manager to reset the chip.
+void noreturn bootstrap_reset(void);
+
+
+// Pseudocode for implementing the public API
+bool  bootstrap_mode_check(void) {
+  uint8_t strap = gpio_straps_read();
+  return strap == BOOTSTRAP_STRAP_VALUE;
+}
+
+rom_error_t bootstrap_mode_enter(void) {
+  const uint8_t jedec_id[] = { ... };
+  bootstrap_jedecid_set(jedec_id);
+
+  sfdp_t sfdp = {
+    // Configure SFDP values according to our internal flash properties.
+  };
+  bootstrap_sfdp_set(&sfdp);
+
+  RETURN_IF_ERROR(bootstrap_phase1());
+  RETURN_IF_ERROR(bootstrap_phase2());
+  bootstrap_reset();
+
+  /* notreached */
+  return kErrorBootstrapImpossibleState;
+}
+```
+
+### Difficulties Anticipated
```

Comment:
As before, this section reads as a proposal rather than what's currently implemented, is that correct?

Reply:
I think this section should stay: as-implemented, the bootstrap module has the described violations of SPI expectations (ie: READ isn't possible, RST does not require RSTEN), and we've not yet tested with the stock dediprog programmer software.

Reply:
Makes sense, thanks